Feat: support sglang backend

vicoooo26 opened the pull request:
Support sglang backend
Ref https://github.com/InftyAI/llmaz/issues/39

kerthcet commented:
/kind feature

kerthcet commented on pkg/controller/inference/playground_controller.go:
```diff
@@ -188,6 +188,13 @@ func buildWorkloadTemplate(model *coreapi.Model, playground *inferenceapi.Playgr
 		"--served-model-name", modelID,
 		"--port", strconv.Itoa(pkg.DEFAULT_BACKEND_PORT),
 	}
+	if bkd.Name() == inferenceapi.SGLANG {
```
I have just refactored this part to make it an interface method, can you change like that:https://github.com/InftyAI/llmaz/pull/47/files#diff-998d843982e36b2a5163d59f67ad4bc159736f103f5a0688bef2ee47b93d8a97R37-R38

kerthcet replied:
To avoid the code conflicts, I'll push the refactors first and cc you once I completed. Then you just need to care about the backend implementations.

vicoooo26 replied:
Already rebased on https://github.com/InftyAI/llmaz/pull/48 and one test case has been added too.

kerthcet reviewed (commented):
Based on the situation that we have no e2e tests, generally LGTM, I'll try to get some test servers. 
Thanks!

kerthcet commented:
/hold

kerthcet commented:
/hold cancel

kerthcet commented on pkg/controller_helper/backend/sglang.go:
```diff
@@ -0,0 +1,75 @@
+	dataSource := datasource.NewDataSourceProvider(model)
+	return []string{
+		"--model-path", dataSource.ModelPath(),
+		"--served-model-name", dataSource.ModelName(),
```
Sorry, does sglang support this flag?

vicoooo26 replied:
https://github.com/sgl-project/sglang/blob/ebf69964cd7e5af9a079eb430ecdd7f67e8566d0/python/sglang/srt/server_args.py#L212
We can find this flag here

kerthcet replied:
That's great

kerthcet commented:
Can you provide an example looks like https://github.com/InftyAI/llmaz/tree/main/docs/examples? If you have no environment, I can help you with that as a follow up.

vicoooo26 commented:
Sure, I will manage to add some docs later

kerthcet commented on docs/examples/huggingface/playground.yaml:
```diff
@@ -5,4 +5,21 @@ metadata:
+        memory: 8Gi
+        cpu: 4
+      limits:
+        nvidia.com/gpu: 1
```
Generally we should not add gpu resource here because the backend didn't need GPU, but the model needs, this is already set in the model config.

vicoooo26 replied:
https://github.com/InftyAI/llmaz/blob/c651c9f8bdcfe5a8410dda8c0768304a14f5ac7d/pkg/controller/inference/service_controller.go#L155 
`Service Controller` will extract a model `InferenceFlavor` info and generate responding `lws` object while `Playground Controller` won't. Maybe I missed something.

kerthcet replied:
see definition here: https://github.com/InftyAI/llmaz/blob/c651c9f8bdcfe5a8410dda8c0768304a14f5ac7d/api/inference/v1alpha1/config_types.go#L44-L48

Generally we'll handle resource requirements at two places, the first place is playground controller, if we set the backendConfig, https://github.com/InftyAI/llmaz/blob/c651c9f8bdcfe5a8410dda8c0768304a14f5ac7d/pkg/controller/inference/playground_controller.go#L201-L209. This is just incase we want to change the default cpu/mem set for backend, e.g. vllm has a default 4C8Gi set, with backendConfig, we can change this dynamically. But for GPU sets, as you pointed out, we'll extract the information from the Model obj in service controller.

So the right place to set the GPUs is at the Model obj, which we'll support fungibility in the future, like serving the llama2-70b, we may have configurations as:

```
inferenceFlavors:
  - name: a100-80gb
    requests:
        nvidia.com/gpu: 2
  - name: a100-40gb
     requests:
        nvidia.com/gpu: 4
```

vicoooo26 replied:
I get it. That will be a great feature. 
XD

kerthcet commented on docs/examples/huggingface/playground.yaml:
```diff
@@ -5,4 +5,21 @@ metadata:
+    args: # additional vllm args
+    - --load-format=auto
+    envs: # optional vllm envs
+    - name: VLLM_USE_MODELSCOPE
```
Minimal configurations is enough here.

vicoooo26 commented:
1. Update `.golangci.yaml` to prevent CI issues
2. Update `README.md` and keep doc dir flat by adding backend prefix for playground to distinguish different backend

BTW I think we could allow user to override backend to their own image registry by adding `DefaultImage` method

kerthcet commented:
> Update .golangci.yaml to prevent CI issues

Thanks @vicoooo26 when reviewing this part, it seems strange to me because I already changed that configurations, but the diff code is still the rotten part, see https://github.com/InftyAI/llmaz/blob/f2085013a435d694a31f1c6a88f679f0f3ad1d02/.golangci.yaml#L20-L23

Any way would not block this PR, can refactor that part later.

> BTW I think we could allow user to override backend to their own image registry by adding DefaultImage method

If you mean providing entrypoint for users to set the image in playground, yes, we can. My original thought is Playground should be as simple as we can. For people who has little knowledge about container, but they know vllm, sglang, tgi very much, they can breezily deploy a model with Playground. However for advanced deployment, they can deploy the Service directly.

At a hight level, Playground is for quick start and easy of use. Service is for extensible, so you can deploy any model with any inference engine as you want. For both deployment approaches, we'll provide scaling, metrics, fungibility support for  them. 

This is the general design about llmaz, who knows, we may change in the future, but this requires feedback from users.

kerthcet commented on pkg/controller_helper/backend/sglang.go:
```diff
@@ -0,0 +1,75 @@
+	sglang_image_registry = "lmsysorg/sglang"
+)
+
+func (v *SGLANG) Name() inferenceapi.BackendName {
```
One nit suggestion:
```
func (s *SGLANG) Name() inferenceapi.BackendName {
```

vicoooo26 replied:
Done

kerthcet reviewed (commented):
One nit, then I'll approve this pr.

vicoooo26 commented:
> > Update .golangci.yaml to prevent CI issues
> 
> Thanks @vicoooo26 when reviewing this part, it seems weird to me because I already changed that configurations, but the diff code is still the rotten part, see
> 
> https://github.com/InftyAI/llmaz/blob/f2085013a435d694a31f1c6a88f679f0f3ad1d02/.golangci.yaml#L20-L23
> 
> Any way would not block this PR, can refactor that part later.
> 
> > BTW I think we could allow user to override backend to their own image registry by adding DefaultImage method
> 
> If you mean providing entrypoint for users to set the image in playground, yes, we can. My original thought is Playground should be as simple as we can. For people who has little knowledge about container, but they know vllm, sglang, tgi very much, they can breezily deploy a model with Playground. However for advanced deployment, they can deploy the Service directly.
> 
> At a hight level, Playground is for quick start and easy of use. Service is for extensible, so you can deploy any model with any inference engine as you want. For both deployment approaches, we'll provide scaling, metrics, fungibility support for them.
> 
> This is the general design about llmaz, who knows, we may change in the future, but this requires feedback from users.

After checking the Github CI action execution history, I found that the previous CI issues happened before PR https://github.com/InftyAI/llmaz/pull/64.  
Since both you and I added this rule, it's enough to keep just one.

kerthcet reviewed (commented):
/lgtm
/approve

Thanks @vicoooo26 !

kerthcet commented:
/lgtm
/approve
